### PR TITLE
Gateway lib unit test cases

### LINKED ISSUE #599: Unit test cases for GatewayLib.sol library
Unit test cases for the following functions.
- test `STAKE_INTENT_TYPEHASH` returns correct hash.
- test `REDEEM_INTENT_TYPEHASH` returns correct hash.
- test `feeAmount()`. This is moved to GatewayBase.sol
- remove `storageVariablePath()` form GatewayLib.sol
- test `proveAccount()` function.
- test `hashStakeIntent ()`function.
- test `hashRedeemIntent()` function.

### PR DESCRIPTION
**Pull request by sarvesh-ost**
Fixes #599 

Unit test cases for GatewayLib.sol

### CODE REVIEW

File: contracts/lib/GatewayLib.sol [hunk 1]
```diff
@@ -29,47 +29,18 @@ library GatewayLib {
 
     /* Constants */
 
-    bytes32 constant STAKE_INTENT_TYPEHASH = keccak256(
+        bytes32 constant public STAKE_INTENT_TYPEHASH = keccak256(
```
Comment (deepesh-kn) [inline contracts/lib/GatewayLib.sol]: Indentation incorrect.

File: contracts/lib/GatewayLib.sol [hunk 2]
```diff
@@ -29,47 +29,18 @@ library GatewayLib {
...
 
-    bytes32 constant REDEEM_INTENT_TYPEHASH = keccak256(
+    bytes32 constant public REDEEM_INTENT_TYPEHASH = keccak256(
         abi.encode(
             "RedeemIntent(uint256 amount,address beneficiary,address gateway)"
         )
     );
 
-
-    /* External Functions */
```
Comment (deepesh-kn) [inline contracts/lib/GatewayLib.sol]: Why deleted this separator comment?

File: contracts/lib/GatewayLib.sol [hunk 3]
```diff
@@ -90,13 +61,16 @@ library GatewayLib {
         pure
         returns (bytes32 storageRoot_)
     {
-        // Decode RLP encoded account value
+        // Decode RLP encoded account value.
```
Comment (deepesh-kn) [inline contracts/lib/GatewayLib.sol]: Fix all the param documentation of `proveAccount` as per the style guide.

File: contracts/lib/GatewayLib.sol [hunk 4]
```diff
@@ -105,8 +79,15 @@ library GatewayLib {
          * Verify the remote OpenST contract against the committed state
```
Comment (deepesh-kn) [inline contracts/lib/GatewayLib.sol]: This should be just a comment and not documentation comment. remove a `*` from line 78

File: contracts/lib/GatewayLib.sol [hunk 5]
```diff
@@ -105,8 +79,15 @@ library GatewayLib {
...
+            MerklePatriciaProof.verify(
+                hashedAccount,
+                _encodedPath,
+                _rlpParentNodes,
+                _stateRoot
+            ),
+            "Account proof is not verified."
+        );
 
         return storageRoot_;
```
Comment (deepesh-kn) [inline contracts/lib/GatewayLib.sol]: Remove explicit return.

File: test/lib/gatewaylib/hash_redeem_intent.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    let expectedRedeemIntent = Utils.hashRedeemIntent(
+      amount,
+      beneficiary,
+      gateway,
+    );
+
+    assert.strictEqual(
+      expectedRedeemIntent,
+      redeemIntent,
+      "Expected redeem intent should be equal to actual.",
```
Comment (deepesh-kn) [inline test/lib/gatewaylib/hash_redeem_intent.js]:
```suggestion
      'Actual redeem intent should be equal to expected redeem intent.',
```

File: test/lib/gatewaylib/hash_stake_intent.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    let expectedStakeIntent = Utils.hashStakeIntent(
+      amount,
+      beneficiary,
+      gateway,
+    );
+
+    assert.strictEqual(
+      expectedStakeIntent,
+      stakeIntent,
+      "Expected stake intent should be equal to actual.",
```
Comment (deepesh-kn) [inline test/lib/gatewaylib/hash_stake_intent.js]:
```suggestion
      'Actual stake intent should be equal to expected stake intent.',
```

File: test/lib/gatewaylib/prove_account.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+//
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+const GatewayLib = artifacts.require('GatewayLib');
+const StubData = require('../../data/proof.json');
+
```
Comment (deepesh-kn) [inline test/lib/gatewaylib/prove_account.js]: remove newline.

File: test/lib/gatewaylib/prove_account.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+      encodedAccountValue,
+      accountProof,
+      path,
+      stateRoot,
+    );
+
+    assert.strictEqual(
+      storageRoot,
+      expectedStorageRoot,
+      'Expected storage root is different from actual.',
```
Comment (deepesh-kn) [inline test/lib/gatewaylib/prove_account.js]:
```suggestion
      'Actual storage root is different from expected storage root.',
```

File: test/lib/gatewaylib/bytes32_to_bytes.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+const GatewayLib = artifacts.require('GatewayLib');
```
Comment (deepesh-kn) [inline test/lib/gatewaylib/bytes32_to_bytes.js]: In `GatewayLib.sol`
1. Why is `bytes32ToBytes` public function and not external?
2. Fix the documentation of `bytes32ToBytes` as per the style guide.
3. Use return variable in `bytes32ToBytes`.

File: contracts/lib/GatewayLib.sol [hunk 11]
```diff
@@ -1,6 +1,6 @@
 pragma solidity ^0.5.0;
 
-// Copyright 2018 OpenST Ltd.
+// Copyright 2019 OpenST Ltd.
```
Comment (deepesh-kn) [inline contracts/lib/GatewayLib.sol]: Add a newline on line number 25.

File: contracts/lib/GatewayLib.sol [hunk 12]
```diff
@@ -29,47 +29,18 @@ library GatewayLib {
...
 
-    bytes32 constant STAKE_INTENT_TYPEHASH = keccak256(
+        bytes32 constant public STAKE_INTENT_TYPEHASH = keccak256(
         abi.encode(
             "StakeIntent(uint256 amount,address beneficiary,address gateway)"
         )
     );
 
-    bytes32 constant REDEEM_INTENT_TYPEHASH = keccak256(
+    bytes32 constant public REDEEM_INTENT_TYPEHASH = keccak256(
```
Comment (deepesh-kn) [inline contracts/lib/GatewayLib.sol]: What's the reason to make both the constants public?

File: contracts/lib/GatewayLib.sol (hunk 11 above)
Comment (deepesh-kn) [reply contracts/lib/GatewayLib.sol]: Why is `import "./RLPEncode.sol";` and `import "./SafeMath.sol";` needed in this library?

File: contracts/lib/GatewayLib.sol (hunk 12 above)
Comment (sarvesh-ost) [reply contracts/lib/GatewayLib.sol]: Adding public creates a getter for these constants.

File: test/lib/gatewaylib/hash_redeem_intent.js (hunk 6 above)
Comment (sarvesh-ost) [reply test/lib/gatewaylib/hash_redeem_intent.js]: why do you like that way? 
It's a way of forming an English sentence. Technically both English sentences are correct.

File: test/lib/gatewaylib/bytes32_to_bytes.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+    let gatewayLib = await GatewayLib.deployed();
+    let bytes32Value = web3.utils.sha3("some data");
+
+    let result = await gatewayLib.bytes32ToBytes(bytes32Value);
+
+    assert.strictEqual(
+      bytes32Value,
+      result,
+      "Expected value should be equal to actual.",
```
Comment (schemar) [inline test/lib/gatewaylib/bytes32_to_bytes.js]: Not sure this tests anything 🤔 Can we somehow assert types with Web3?

File: test/lib/gatewaylib/hash_redeem_intent.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+// ----------------------------------------------------------------------------
+
+const GatewayLib = artifacts.require('GatewayLib');
+
+const Utils = require('../../gateway/eip20_cogateway/helpers/co_gateway_utils');
+const BN = require('bn.js');
+
+contract('GatewayLib.hashRedeemIntent()', async (accounts) => {
+
+  it('should return correct redeem intent', async function () {
```
Comment (schemar) [inline test/lib/gatewaylib/hash_redeem_intent.js]:
```suggestion
  it('should return correct redeem intent hash', async function () {
```

File: test/lib/gatewaylib/hash_redeem_intent.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+    let gatewayLib = await GatewayLib.deployed();
+    let amount = new BN(1);
+    let beneficiary = accounts[0];
+    let gateway = accounts[1];
+
+    let redeemIntent = await gatewayLib.hashRedeemIntent(
```
Comment (schemar) [inline test/lib/gatewaylib/hash_redeem_intent.js]:
```suggestion
    let redeemIntentHash = await gatewayLib.hashRedeemIntent(
```

File: test/lib/gatewaylib/hash_redeem_intent.js [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+      amount,
+      beneficiary,
+      gateway,
+    );
+
+    let expectedRedeemIntent = Utils.hashRedeemIntent(
```
Comment (schemar) [inline test/lib/gatewaylib/hash_redeem_intent.js]:
```suggestion
    let expectedRedeemIntentHash = Utils.hashRedeemIntent(
```

File: test/lib/gatewaylib/hash_redeem_intent.js (hunk 6 above)
Comment (schemar) [reply test/lib/gatewaylib/hash_redeem_intent.js]: How about `'The library did not hash the redeem intent as expected'`?

File: test/lib/gatewaylib/hash_stake_intent.js [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+// ----------------------------------------------------------------------------
+
+const GatewayLib = artifacts.require('GatewayLib');
+
+const Utils = require('../../gateway/eip20_cogateway/helpers/co_gateway_utils');
+const BN = require('bn.js');
+
+contract('GatewayLib.hashStakeIntent()', async (accounts) => {
+
+  it('should return correct stake intent', async function () {
```
Comment (schemar) [inline test/lib/gatewaylib/hash_stake_intent.js]:
```suggestion
  it('should return correct stake intent hash', async function () {
```

File: test/lib/gatewaylib/hash_stake_intent.js [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+    let gatewayLib = await GatewayLib.deployed();
+    let amount = new BN(1);
+    let beneficiary = accounts[0];
+    let gateway = accounts[1];
+
+    let stakeIntent = await gatewayLib.hashStakeIntent(
```
Comment (schemar) [inline test/lib/gatewaylib/hash_stake_intent.js]:
```suggestion
    let stakeIntentHash = await gatewayLib.hashStakeIntent(
```

File: test/lib/gatewaylib/hash_stake_intent.js [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+      amount,
+      beneficiary,
+      gateway,
+    );
+
+    let expectedStakeIntent = Utils.hashStakeIntent(
```
Comment (schemar) [inline test/lib/gatewaylib/hash_stake_intent.js]:
```suggestion
    let expectedStakeIntentHash = Utils.hashStakeIntent(
```

File: test/lib/gatewaylib/hash_stake_intent.js (hunk 7 above)
Comment (schemar) [reply test/lib/gatewaylib/hash_stake_intent.js]: Same as above?

File: test/lib/gatewaylib/prove_account.js [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+    );
+  });
+
+  it('should fail for wrong storage root', async function () {
```
Comment (schemar) [inline test/lib/gatewaylib/prove_account.js]: Wrong storage root or wrong state root?

File: test/lib/gatewaylib/prove_account.js (hunk 8 above)
Comment (sarvesh-ost) [reply test/lib/gatewaylib/prove_account.js]: Check latest formating.

File: test/lib/gatewaylib/bytes32_to_bytes.js (hunk 13 above)
Comment (sarvesh-ost) [reply test/lib/gatewaylib/bytes32_to_bytes.js]: I don't think, we can assert types in truffle. This method only changes type not value, an assertion that the same value is returned seems valid  🤔  

Any suggestions?

Comment (schemar) [reply test/lib/gatewaylib/bytes32_to_bytes.js]: No other suggestion. If Web3 cannot be used to assert types, then there seems nothing else we can do.

**Pull request merged**